File: 079_word_search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        if (board.length == 0 || board[0].length == 0) return false;
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (dfs(board, 0, word, i, j)) return true;
            }
        }
        return false;
    }

    private boolean dfs(char[][] board, int wordIdx, String word, int i, int j) {
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length) {
            return false;
        }

        boolean result = false;
        if (board[i][j] == word.charAt(wordIdx)) {
            wordIdx++;
            if (wordIdx == word.length()) {
                return true;
            }
            char temp = board[i][j];
            board[i][j] = '#';
            result = dfs(board, wordIdx, word, i - 1, j) || dfs(board, wordIdx, word, i + 1, j) ||
                    dfs(board, wordIdx, word, i, j - 1) || dfs(board, wordIdx, word, i, j + 1);
            board[i][j] = temp;
        }
        return result;
    }
}

File: python/079_word_search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def dfs(board: List[List[str]], i: int, j: int, word: str) -> bool:
            if len(word) == 0: return True
            if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or board[i][j] != word[0]:
                return False
            res = False
            board[i][j] = '#'
            for dx, dy in [[-1, 0], [0, 1], [1, 0], [0, -1]]:
                x, y = i + dx, j + dy
                res = dfs(board, x, y, word[1:])
                if res: break
            board[i][j] = word[0]
            return res

        if not board or not board[0]: return False
        for i in range(len(board)):
            for j in range(len(board[0])):
                if dfs(board, i, j, word):
                    return True
        return False
